### ISSUE TITLE
Reflections method numFields does not count parent fields

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#8736]
While playing around with [Reflections](https://chapel-lang.org/docs/latest/modules/standard/Reflection.html?highlight=reflection) to try to ghetto a solution to [testing](https://github.com/chapel-lang/chapel/issues/8724) I noted this.

```chapel
use Reflection;

class TestCase {

  var tests : domain(string);
  var n: string = "fred";

  proc init() { }

  proc run() {
    for test in tests {
      writeln("running test ", test);
    }
  }
}

class ChicoTest : TestCase {
  proc init() {
    super.init();
  }

  proc testBobbyLake() {
    writeln("Bobby Lake is a good lake");
  }
}

var ct = new ChicoTest();
ct.tests.add("testBobbyLake");
writeln(numFields(ChicoTest));
```

produces 0.  But if you move `var n: string = "fred";` into `ChicoTest`, it produces 1.  Is that expected?  It seems to me that `ChicoTest` has two fields.

**Comment by daviditen** [comment]
The `numFields` procedure currently only returns the number of fields directly defined by the given class not including inherited fields.  I think it's reasonable to want to know both that and the number of fields including inherited fields.  The procedure below returns the number of fields defined by the given class including inherited fields.

```
proc numAllFields(type t) param {
 if t == object {
   return numFields(t);
 } else {
   var x:t;
   return numFields(t) + numAllFields(x.super.type);
 }
}
```

**Comment by buddha314** [comment]
Oooh, that's neat!  Should that go into Reflections or is it too niche?

**Comment by daviditen** [comment]
IMO yes, something like this should go into `Reflections` with a name that is probably not `numAllFields`.

**Comment by buddha314** [comment]
Should I re-name this ticket or something?  And don't ask me for a name, because I would choose `whatAllUpInThisBusiness()`

**Comment by daviditen** [comment]
Maybe change the title to "Support counting inherited fields in Reflection.numFields()"?  With a bit more time to think about it, I think the ideal way would be to have `numFields` take an argument to specify whether or not to count inherited fields.

**Comment by bradcray** [comment]
Thanks for jumping on this one @daviditen : I like the idea of the default arg indicating whether or not to worry about inheritance (my mind went the same way).  A challenge is what to do about all of the other `getField*()` methods which take an index number of the field...  Somehow they'd need to know whether to include parent fields or not as well.  Maybe this is just a matter of having them take a bool as well, but it leads to questions like "What order are the fields enumerated in?"  (where my gut answer would be "if including the parent classes, start at the top of the hierarchy."  I think that's natural, though it also means that field 'i' would have a different definition depending on the value of the bool...

**Comment by dlongnecke-cray** [comment]
Bumping this as it relates to the Reflection module review.

As I see it there are three positions we could take here:

- Keep the current behavior of `numFields()`, and only return the number of fields in the current class
- Have `numFields()` include the number of fields in all superclasses (@daviditen function above)
- Use some sort of flag to indicate the behavior

Where for option 3, the question is should the flag have a default, and what should the default be?

---

My position is that `numFields` should report the fields in the statically derived class, and not any parents. We could have a flag, e.g. `countSuperFields`, that is by default false. If set to true, it would enable a behavior similar to David's function above.

**Comment by cassella** [comment]
Having `numFields()` by default return anything other than the number of fields the type has seems extremely unintuitive.  I'd say a derived class does "have" the fields it inherits from its parent.  

I'm not sure the current behavior even needs to be provided as an option.  Are there cases where you only want to look at the non-inherited fields?  If there are, you could do something corresponding to diten's example above, and know they're `numFields(x.super.type)..<numFields(t)`.  (Where `numFields()` here is the inclusive interpretation.)

Two other flavors of the function might be:

Should it include fields that the type appears to have due to forwarding?  It doesn't today.  I don't have a good intuition for whether it should.  It doesn't "have" those fields, but those fields can be accessed.  I'd lean towards no due to the first point.

When private fields are supported, should it include those?  I think it should, since it has those fields, and there should be a `Reflection` method to find out if a field at a given index is private.

**Comment by bradcray** [comment]
I like Paul's argument (but am open to counterarguments too).

> Should it include fields that the type appears to have due to forwarding? It doesn't today. I don't have a good intuition for whether it should. It doesn't "have" those fields, but those fields can be accessed. I'd lean towards no due to the first point.

My intuition is the same.  I think we could also rationalize not including them by saying that what's really forwarded is a paren-less method that accesses the forwarded field (though... maybe I'm just trying to create rationalizations).

**Comment by dlongnecke-cray** [comment]
This is probably very subjective, but when I look at the fields of a class, I almost never consider the fields of superclasses to be part of the class, at least not in the sense that I think of them as part of the "logical interface" of the class. Obviously, they're stored in the same chunk of memory in the end, but I don't reason about them in the same way as a field declared within the derived class itself. This might be related to how I tend to view encapsulation.

Having said that, I think it would be useful to easily be able to access superclass fields, but were that the default I'd find it very confusing given how I tend to view class memory.

I'm still in favor of the bool, personally, but I'd be fine with Paul's interpretation even if it would take me a little getting used to.

I wasn't aware that forwarding gave you direct access to the fields of a forwarded field. I don't think it should, that seems very weird to me. I like Brad's rationalization!

**Comment by bradcray** [comment]
> when I look at the fields of a class, I almost never consider the fields of superclasses to be part of the class

Is that true in all cases, David, or does it vary from class to class?  For example, in a simple OOP 101 case, like:

```chapel
class Point {
  var x, y: real;
}

class ColoredPoint: Point {
  var c: color;
}

class 3DPoint: Point {
  var z: real;
}
```

it seems difficult for me to imagine that you wouldn't consider x and y to be fields of ColoredPoint or 3DPoint if you were inspecting them, creating a generic print routine for them, etc.  Do you not?

**Comment by dlongnecke-cray** [comment]
I guess I would expect to access `x` and `y` via `x()` and `y()` (or their parenless equivalents) instead of via direct accesses. And whether or not those fields are writeable would depend on the purpose of the type. I think it's more a style thing than anything else, but I kind of consider mutating superclass fields as "none of the child class's business".

This just comes down to my view on interface design more than anything else. I think either way we go on this question, we will have the capability to produce the other behavior.

At the very least I would expect both options to be possible (and I don't think anyone else here has suggested otherwise, either).

**Comment by arezaii** [comment]
Following up from module review with a note about how Java handles a similar issue in regards to super fields. The summary is that Java Reflection provides two named methods, one for getting the fields/methods of the class only, and one that contains those of the superclass as well. While `getFields()` includes those from the superclass, `getDeclaredFields()` only provides those declared in the class.

https://www.oracle.com/technical-resources/articles/java/javareflection.html

Not saying it's the best way, but it seems a reasonable and at least somewhat intuitive pattern to follow.

For comparison sake, I looked briefly at Swift which implements this functionality in their `Mirror` API. The `children` property contains all the elements belonging to that class. If you want the parent's elements, there is a `superclassMirror` property that you can use to walk up the inheritance tree and identify elements belonging to each superclass.

https://developer.apple.com/documentation/swift/mirror

**Comment by bradcray** [comment]
I know a lot of things were said during the meeting today, but my takeaways were:

* different people have different expectations about what should happen here (within the team, and between the team and users), so I feel increasingly confident we should support the ability to count inherited fields and not
* I haven't heard anyone express support for having forwarding come into play here at all.
* we talked a lot about having a boolean toggle for this in the meeting, but given that:
  * there wasn't much clarity on what default it should have
  * there wasn't much support for having it not default
  * boolean flags tend to either require pass-by-keyword or else are vague
  * Java does this through two routines
  ...I'm increasingly thinking we should do it through two routines as well

So that would suggest the main challenge is naming (as usual), and whether we change numFields() to include the superclass (which the Java naming might suggest) or come up with a new name for that case (like `numAllFields()`, `totalNumFields()`?).  As Michael and Lydia noted, we'll need two forms of `getField(i: int)` or using Engin's naming `fieldAtIndex(param i: int)`

> (or their parenless equivalents) instead of via direct accesses.

David, how can you distinguish between a direct access to `x` and a parenless access to it through a method named `x`?

**Comment by dlongnecke-cray** [comment]
> David, how can you distinguish between a direct access to x and a parenless access to it through a method named x?

Yeah, took me minute to realize that in Chapel you really can't. Or at least the user can't, but the compiler can :D.

So yeah, the compiler does auto generate accessors, but I think it still might come down to intention more than anything else (and in a world with private methods those accessors might be marked private).
